File: android/airship-framework-proxy/src/test/java/com/urbanairship/android/framework/proxy/TagOperationTest.kt
package com.urbanairship.android.framework.proxy

import com.urbanairship.channel.TagEditor
import com.urbanairship.json.JsonException
import com.urbanairship.json.JsonValue
import io.mockk.confirmVerified
import io.mockk.mockk
import io.mockk.verify
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
public class TagOperationTest {

    @Test
    public fun testParseAdd() {
        val json = """
            {
                "tags": ["oneTag", "anotherTag", "andALastOne"],
                "operationType": "add"
            }
        """

        val parsed = TagOperation(json = JsonValue.parseString(json).requireMap())

        val expected = TagOperation(
            tags = listOf("oneTag", "anotherTag", "andALastOne"),
            action = TagOperationAction.ADD
        )

        Assert.assertEquals(expected, parsed)
    }

    @Test
    public fun testParseRemove() {
        val json = """
            {
                "tags": ["oneTag", "anotherTag", "andALastOne"],
                "operationType": "remove"
            }
        """

        val parsed = TagOperation(json = JsonValue.parseString(json).requireMap())

        val expected = TagOperation(
            tags = listOf("oneTag", "anotherTag", "andALastOne"),
            action = TagOperationAction.REMOVE
        )

        Assert.assertEquals(expected, parsed)
    }

    @Test(expected = IllegalArgumentException::class)
    public fun testParseInvalidAction() {
        val json = """
            {
                "tags": ["oneTag", "anotherTag", "andALastOne"],
                "operationType": "invalidAction"
            }
        """

        TagOperation(json = JsonValue.parseString(json).requireMap())
    }

    @Test(expected = JsonException::class)
    public fun testParseInvalidTags() {
        val json = """
            {
                "tags": "invalidTag",
                "operationType": "add"
            }
        """

        TagOperation(json = JsonValue.parseString(json).requireMap())
    }

    @Test
    public fun testApplyAdd() {
        val editor = mockk<TagEditor>(relaxed = true)

        val tags = listOf("oneTag", "anotherTag", "andALastOne")

        val operation = TagOperation(
            tags = tags,
            action = TagOperationAction.ADD
        )

        operation.applyOperation(editor)
        verify { editor.addTags(tags.toSet()) }
        confirmVerified(editor)
    }

    @Test
    public fun testApplyRemove() {
        val editor = mockk<TagEditor>(relaxed = true)

        val tags = listOf("oneTag", "anotherTag", "andALastOne")

        val operation = TagOperation(
            tags = tags,
            action = TagOperationAction.REMOVE
        )

        operation.applyOperation(editor)
        verify { editor.removeTags(tags.toSet()) }
        confirmVerified(editor)
    }

}

File: ios/AirshipFrameworkProxyTests/TagOperationTest.swift
import XCTest

@testable import AirshipFrameworkProxy
import AirshipKit

class TagOperationTest: XCTestCase {

    func testParseAdd() throws {
        let json = """
            {
                "tags": ["oneTag", "anotherTag", "andALastOne"],
                "operationType": "add"
            }
        """

        let parsed = try JSONDecoder().decode(TagOperation.self, from: json.data(using: .utf8)!)

        let expected = TagOperation(
            action: TagOperation.Action.addTags,
            tags: ["oneTag", "anotherTag", "andALastOne"]
        )

        XCTAssertEqual(expected, parsed)
    }
    
    func testParseRemove() throws {
        let json = """
            {
                "tags": ["oneTag", "anotherTag", "andALastOne"],
                "operationType": "remove"
            }
        """

        let parsed = try JSONDecoder().decode(TagOperation.self, from: json.data(using: .utf8)!)

        let expected = TagOperation(
            action: TagOperation.Action.removeTags,
            tags: ["oneTag", "anotherTag", "andALastOne"]
        )

        XCTAssertEqual(expected, parsed)
    }

    func testParseInvalidAction() throws {
        let json = """
           {
                "tags": ["oneTag", "anotherTag", "andALastOne"],
                "operationType": "invalid"
           }
        """

        do {
            let _ = try JSONDecoder().decode(TagOperation.self, from: json.data(using: .utf8)!)
            XCTFail("should throw")
        } catch {}
    }
    
    func testParseInvalidTags() throws {
        let json = """
           {
                "tags": "invalid",
                "operationType": "add"
           }
        """

        do {
            let _ = try JSONDecoder().decode(TagOperation.self, from: json.data(using: .utf8)!)
            XCTFail("should throw")
        } catch {}
    }
    
    func testApply() throws {
        let editor = TestEditor()
        let operation = TagOperation(
            action: TagOperation.Action.addTags,
            tags: ["oneTag", "anotherTag", "andALastOne"]
        )
        operation.apply(editor: editor)
        XCTAssertEqual(editor.currentTags, ["oneTag", "anotherTag", "andALastOne"])
        
        let nextOperation = TagOperation(
            action: TagOperation.Action.removeTags,
            tags: ["anotherTag", "andALastOne"]
        )
        nextOperation.apply(editor: editor)
        XCTAssertEqual(editor.currentTags, ["oneTag"])
    }

}

fileprivate class TestEditor: TagOperationEditor {
    var currentTags: Set<String> = Set()

    func add(_ tags: [String]) {
        for tag in tags {
            currentTags.insert(tag)
        }
    }

    func remove(_ tags: [String]) {
        for tag in tags {
            currentTags.remove(tag)
        }
    }
}
